### PR TITLE
GDB sim

### PR DESCRIPTION
**Pull request by kito-cheng**
This PR contain E, C, F and D extension and several  gdb sim fix.

### CODE REVIEW

File: sim/riscv/sim-main.c [hunk 1]
```diff
@@ -1451,7 +1451,7 @@ execute_i (SIM_CPU *cpu, unsigned_word iw, const struct riscv_opcode *op)
       store_rd (cpu, rd, EXTEND32 (ashiftrt ((signed32)cpu->regs[rs1], cpu->regs[rs2] & 0x1f)));
       break;
     case MATCH_SRAI:
-      TRACE_INSN (cpu, "srai %s, %s, %"PRIiTW";  // %s = %s >>> %#"PRIxTW,
+      TRACE_INSN (cpu, "srai %s, %s, %" PRIiTW ";  // %s = %s >>> %#"PRIxTW,
```
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: You put spaces around the first PRIiTW but not the second one.  This is a minor nit.

File: sim/riscv/sim-main.c [hunk 2]
```diff
@@ -1467,23 +1467,30 @@ execute_i (SIM_CPU *cpu, unsigned_word iw, const struct riscv_opcode *op)
...
 		  rd_name, rs1_name, shamt_imm, rd_name, rs1_name, shamt_imm);
       RISCV_ASSERT_RV64 (cpu, "insn: %s", op->name);
-      store_rd (cpu, rd, EXTEND32 (ashiftrt ((signed32)cpu->regs[rs1], shamt_imm)));
+      store_rd (cpu, rd,
+		EXTEND32 (ashiftrt ((signed32)cpu->regs[rs1], shamt_imm)));
       break;
     case MATCH_SLT:
-      TRACE_INSN (cpu, "slt");
-      store_rd (cpu, rd, !!((signed_word)cpu->regs[rs1] < (signed_word)cpu->regs[rs2]));
+      TRACE_INSN (cpu, "slt %s %s %s", rd_name, rs1_name, rs2_name);
```
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: I would expect commas between the instruction operands.  This is a minor nit.

File: sim/riscv/sim-main.c [hunk 3]
```diff
@@ -1286,13 +1286,19 @@ execute_i (SIM_CPU *cpu, unsigned_word iw, const struct riscv_opcode *op)
   unsigned_word sb_imm = EXTRACT_SBTYPE_IMM (iw);
   unsigned_word shamt_imm = ((iw >> OP_SH_SHAMT) & OP_MASK_SHAMT);
   unsigned_word tmp;
+  unsigned_word sys_id;
+  int eh_rve_p = cpu->elf_flags & 0x8;
```
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: Should use an EF_* value from include/elf/riscv.h, except that one hasn't been defined for RVE.  So probably should define one first.

File: sim/riscv/sim-main.c [hunk 4]
```diff
@@ -1653,23 +2124,20 @@ void step_once (SIM_CPU *cpu)
...
-      sim_io_printf (sd, "sim: bad insn len %#x @ %#"PRIxTA": %#"PRIxTW"\n",
-		     len, pc, iw);
-      sim_engine_halt (sd, cpu, NULL, pc, sim_signalled, SIM_SIGILL);
-    }
 
-  iw |= ((unsigned_word)sim_core_read_aligned_2 (cpu, pc, exec_map, pc + 2) << 16);
+  if (len == 4)
+    iw |= ((unsigned_word)sim_core_read_aligned_2 (cpu, pc, exec_map, pc + 2) << 16);
+  else
+    iw |= ((unsigned_word)sim_core_read_aligned_2 (cpu, pc, exec_map, pc));
```
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: This looks redundant.  For a 2-byte instruction, we already read two bytes.  No need to re-read the same 2 bytes again.

File: sim/riscv/sim-main.c [hunk 5]
```diff
@@ -1678,11 +2146,26 @@ void step_once (SIM_CPU *cpu)
...
+      if ((op->pinfo & INSN_ALIAS))
+	continue;
+      /* Is this instruction restricted to a certain value of XLEN?  */
+      if (isdigit (op->subset[0]) && atoi (op->subset) != xlen)
+	continue;
+      /* It's a match.  */
+      pc = execute_one (cpu, iw, op);
+      break;
+
+      if ((op->match_func) (op, iw) && !(op->pinfo & INSN_ALIAS))
```
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: Dead code after a break that should be removed.  This is the original code which is no longer useful.

File: sim/riscv/interp.c [hunk 6]
```diff
@@ -154,13 +158,23 @@ sim_create_inferior (SIM_DESC sd, struct bfd *abfd,
...
+  Elf_Internal_Phdr *phdr;
+  int i, phnum;
 
   /* Set the PC.  */
   if (abfd != NULL)
     addr = bfd_get_start_address (abfd);
   else
     addr = 0;
   sim_pc_set (cpu, addr);
+  phdr = elf_tdata (abfd)->phdr;
```
Comment (jim-wilson) [inline sim/riscv/interp.c]: This assumes the highest address in the executable is the start of heap.  But if you use a linker script, you could have put special sections in high memory, and putting heap after them would be wrong.  Or alternatively, a board could have multiple memory regiions, and the region where the text/data is stored is not necessarily the highest address one.  Normally, one looks for the symbol _end and uses that as the start of heap.  But there is no one right way to do this.

File: sim/riscv/sim-main.c [hunk 7]
```diff
@@ -817,6 +818,21 @@ execute_i (SIM_CPU *cpu, unsigned_word iw, const struct riscv_opcode *op)
   unsigned_word shamt_imm = ((iw >> OP_SH_SHAMT) & OP_MASK_SHAMT);
   unsigned_word tmp;
   sim_cia pc = cpu->pc + 4;
+  host_callback *cb = STATE_CALLBACK (sd);
```
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: This block of code should go inside the MATCH_ECALL case where is the only place it is used.  It is wasted work if we don't have a syscall.  You can add a brace level for the case MATCH_ECALL for the new local variables.

File: sim/riscv/sim-main.c (hunk 1 above)
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: You put spaces around the first PRIiTW but not the second one. This is a minor nit.

File: sim/riscv/sim-main.c (hunk 2 above)
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: I would expect commas between the instruction operands. This is a minor nit. This occurs 4 times.

File: sim/riscv/sim-main.c (hunk 7 above)
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: This block of code should go inside the MATCH_ECALL case where is the only place it is used. It is wasted work if we don't have a syscall. You can add a brace level for the case MATCH_ECALL for the new local variables.

File: sim/riscv/interp.c (hunk 6 above)
Comment (jim-wilson) [inline sim/riscv/interp.c]: This assumes the highest address in the executable is the start of heap. But if you use a linker script, you could have put special sections in high memory, and putting heap after them would be wrong. Or alternatively, a board could have multiple memory regiions, and the region where the text/data is stored is not necessarily the highest address one. Normally, one looks for the symbol _end and uses that as the start of heap. But there is no one right way to do this.

File: sim/riscv/sim-main.c (hunk 4 above)
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: This looks redundant. For a 2-byte instruction, we already read two bytes. No need to re-read the same 2 bytes again.

File: sim/riscv/sim-main.c (hunk 5 above)
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: Dead code after a break that should be removed. This is the original code which is no longer useful.

File: sim/riscv/sim-main.c (hunk 3 above)
Comment (jim-wilson) [inline sim/riscv/sim-main.c]: Should use an EF_* value from include/elf/riscv.h, except that one hasn't been defined for RVE. So probably should define one first.

Comment (kito-cheng) [reply sim/riscv/sim-main.c]: Yeah, I think the problem is RVE support not merge to riscv-binutils-2.29 yet, I'll remove RVE in this PR and send another for this.

File: sim/riscv/interp.c (hunk 6 above)
Comment (kito-cheng) [reply sim/riscv/interp.c]: I've update this part, it will use `_end` if possible.

File: sim/riscv/sim-main.c (hunk 1 above)
Comment (kito-cheng) [reply sim/riscv/sim-main.c]: Fixed.

File: sim/riscv/sim-main.c (hunk 2 above)
Comment (kito-cheng) [reply sim/riscv/sim-main.c]: Fixed.

File: sim/riscv/sim-main.c (hunk 4 above)
Comment (kito-cheng) [reply sim/riscv/sim-main.c]: Fixed.

File: sim/riscv/sim-main.c (hunk 5 above)
Comment (kito-cheng) [reply sim/riscv/sim-main.c]: Fixed.

File: sim/riscv/sim-main.c (hunk 7 above)
Comment (kito-cheng) [reply sim/riscv/sim-main.c]: Fixed.

**Comment by jim-wilson** [comment]
I am unfamiliar with github workflow in general, and riscv github workflow in particular.  I was assuming that the person who submitted the patch would merge them after approval.  This is how FSF binutils/gcc work.  If you are waiting for me to merge them, let me know.

**Comment by jim-wilson** [comment]
I see that this is a request to add patches to riscv-binutils-2.29.  Should this be a request to add patches to riscv-next instead?  And then maybe cherry pick onto the riscv-binutils-2.29 branch?  There needs to be some clearly defined process here, or otherwise we end up with two different simulator ports for no reason.

**Comment by palmer-dabbelt** [comment]
The problem with PRs against riscv-next is that github breaks them as riscv-next gets rebased.  I've been manually handling the impedance mismatch, but I think we might want to change our workflow a bit and do away with the "riscv-next" branch entirely.  I'm at the workshop, but let's talk when I'm back.

**Comment by jim-wilson** [comment]
I was hoping for info from Palmer on how to manage riscv branches, but he has been busy, so I went ahead and merged this, and will deal with the fallout later.

**Pull request merged**